xbmc/network/upnp/UPnPInternal.cpp

```diff
@@ -1150,22 +1174,24 @@ bool GetResource(const PLT_MediaObject* entry, CFileItem& item)
   }
 
   // look for subtitles
-  unsigned subs = 0;
+  unsigned subIdx = 0;
+
   for(unsigned r = 0; r < entry->m_Resources.GetItemCount(); r++)
   {
     const PLT_MediaItemResource& res  = entry->m_Resources[r];
     const PLT_ProtocolInfo&      info = res.m_ProtocolInfo;
-    static const char* allowed[] = { "text/srt"
-      , "text/ssa"
-      , "text/sub"
-      , "text/idx" };
-    for(const char* const type : allowed)
+
+    for (const std::string& type : ALLOWED_REAL_SUBTITLE_FORMAT)
     {
-      if(info.Match(PLT_ProtocolInfo("*", "*", type, "*")))
+      if (type == (std::string)info.GetContentType().Split("/").GetLastItem()->GetChars())
       {
-        std::string prop = StringUtils::Format("subtitle:{}", ++subs);
+        ++subIdx;
+        logger->info("adding subtitle: #{}, type '{}', URI '{}'", subIdx, type,
+                     (const char*)res.m_Uri);
```

Comment:
```suggestion
                     res.m_Uri);
```

Please drop the cast, fmt is now able to cope with it just fine

Reply:
I tried,if I remove the cast,build will fail

```
[ 96%] Building CXX object build/network/upnp/CMakeFiles/network_upnp.dir/UPnPInternal.cpp.o
[ 97%] Built target network_httprequesthandlers
[100%] Built target legacy_interface
In file included from /usr/include/fmt/format.h:48,
                 from /home/ihipop/.cache/yay/kodi-git/src/xbmc/xbmc/utils/StringUtils.h:37,
                 from /home/ihipop/.cache/yay/kodi-git/src/xbmc/xbmc/utils/Digest.h:11,
                 from /home/ihipop/.cache/yay/kodi-git/src/xbmc/xbmc/Util.h:12,
                 from /home/ihipop/.cache/yay/kodi-git/src/xbmc/xbmc/network/upnp/UPnPInternal.cpp:16:
/usr/include/fmt/core.h: In substitution of ‘template<class T, class Context> using mapped_type_constant = fmt::v9::detail::type_constant<decltype (fmt::v9::detail::arg_mapper<Context>().map(declval<const T&>())), typename Context::char_type> [with T = NPT_String; Context = fmt::v9::basic_format_context<fmt::v9::appender, char>]’:
/usr/include/fmt/core.h:1730:41:   recursively required from ‘constexpr long long unsigned int fmt::v9::detail::encode_types() [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; Arg = std::basic_string_view<char>; Args = {NPT_String}]’
/usr/include/fmt/core.h:1730:41:   required from ‘constexpr long long unsigned int fmt::v9::detail::encode_types() [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; Arg = unsigned int; Args = {std::basic_string_view<char, std::char_traits<char> >, NPT_String}]’
/usr/include/fmt/core.h:1886:58:   required from ‘constexpr const long long unsigned int fmt::v9::format_arg_store<fmt::v9::basic_format_context<fmt::v9::appender, char>, unsigned int, std::basic_string_view<char, std::char_traits<char> >, NPT_String>::desc’
/usr/include/fmt/core.h:1999:63:   required from ‘constexpr fmt::v9::basic_format_args<Context>::basic_format_args(const fmt::v9::format_arg_store<Context, Args ...>&) [with Args = {unsigned int, std::basic_string_view<char, std::char_traits<char> >, NPT_String}; Context = fmt::v9::basic_format_context<fmt::v9::appender, char>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:370:36:   required from ‘void spdlog::logger::log_(spdlog::source_loc, spdlog::level::level_enum, spdlog::string_view_t, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; spdlog::string_view_t = fmt::v9::basic_string_view<char>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:90:13:   required from ‘void spdlog::logger::log(spdlog::source_loc, spdlog::level::level_enum, fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:96:12:   required from ‘void spdlog::logger::log(spdlog::level::level_enum, fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:158:12:   required from ‘void spdlog::logger::info(fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/xbmc/xbmc/network/upnp/UPnPInternal.cpp:1187:21:   required from here
/usr/include/fmt/core.h:1530:53: 错误：对重载的‘map(const NPT_String&)’的调用有歧义
 1530 |     type_constant<decltype(arg_mapper<Context>().map(std::declval<const T&>())),
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/fmt/core.h:1373:33: 附注：candidate: ‘constexpr bool fmt::v9::detail::arg_mapper<Context>::map(bool) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>]’
 1373 |   FMT_CONSTEXPR FMT_INLINE auto map(bool val) -> bool { return val; }
      |                                 ^~~
/usr/include/fmt/core.h:1398:33: 附注：candidate: ‘constexpr const fmt::v9::detail::arg_mapper<Context>::char_type* fmt::v9::detail::arg_mapper<Context>::map(char_type*) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; char_type = char]’
 1398 |   FMT_CONSTEXPR FMT_INLINE auto map(char_type* val) -> const char_type* {
      |                                 ^~~
/usr/include/fmt/core.h:1401:33: 附注：candidate: ‘constexpr const fmt::v9::detail::arg_mapper<Context>::char_type* fmt::v9::detail::arg_mapper<Context>::map(const char_type*) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; char_type = char]’
 1401 |   FMT_CONSTEXPR FMT_INLINE auto map(const char_type* val) -> const char_type* {
      |                                 ^~~
/usr/include/fmt/core.h:1437:33: 附注：candidate: ‘constexpr const void* fmt::v9::detail::arg_mapper<Context>::map(void*) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>]’
 1437 |   FMT_CONSTEXPR FMT_INLINE auto map(void* val) -> const void* { return val; }
      |                                 ^~~
/usr/include/fmt/core.h:1438:33: 附注：candidate: ‘constexpr const void* fmt::v9::detail::arg_mapper<Context>::map(const void*) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>]’
 1438 |   FMT_CONSTEXPR FMT_INLINE auto map(const void* val) -> const void* {
      |                                 ^~~
/usr/include/fmt/core.h:1524:8: 附注：candidate: ‘fmt::v9::detail::unformattable fmt::v9::detail::arg_mapper<Context>::map(...) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>]’
 1524 |   auto map(...) -> unformattable { return {}; }
      |        ^~~
/usr/include/fmt/core.h: In instantiation of ‘constexpr const long long unsigned int fmt::v9::format_arg_store<fmt::v9::basic_format_context<fmt::v9::appender, char>, unsigned int, std::basic_string_view<char, std::char_traits<char> >, NPT_String>::desc’:
/usr/include/fmt/core.h:1999:63:   required from ‘constexpr fmt::v9::basic_format_args<Context>::basic_format_args(const fmt::v9::format_arg_store<Context, Args ...>&) [with Args = {unsigned int, std::basic_string_view<char, std::char_traits<char> >, NPT_String}; Context = fmt::v9::basic_format_context<fmt::v9::appender, char>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:370:36:   required from ‘void spdlog::logger::log_(spdlog::source_loc, spdlog::level::level_enum, spdlog::string_view_t, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; spdlog::string_view_t = fmt::v9::basic_string_view<char>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:90:13:   required from ‘void spdlog::logger::log(spdlog::source_loc, spdlog::level::level_enum, fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:96:12:   required from ‘void spdlog::logger::log(spdlog::level::level_enum, fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:158:12:   required from ‘void spdlog::logger::info(fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/xbmc/xbmc/network/upnp/UPnPInternal.cpp:1187:21:   required from here
/usr/include/fmt/core.h:1886:58:   in ‘constexpr’ expansion of ‘fmt::v9::detail::encode_types<fmt::v9::basic_format_context<fmt::v9::appender, char>, unsigned int, std::basic_string_view<char, std::char_traits<char> >, NPT_String>()’
/usr/include/fmt/core.h:1730:41:   in ‘constexpr’ expansion of ‘fmt::v9::detail::encode_types<fmt::v9::basic_format_context<fmt::v9::appender, char>, std::basic_string_view<char>, NPT_String>()’
/usr/include/fmt/core.h:1730:41:   in ‘constexpr’ expansion of ‘fmt::v9::detail::encode_types<fmt::v9::basic_format_context<fmt::v9::appender, char>, NPT_String>()’
/usr/include/fmt/core.h:1885:39: 错误：‘constexpr’ call flows off the end of the function
 1885 |   static constexpr unsigned long long desc =
      |                                       ^~~~
/usr/include/fmt/core.h: In instantiation of ‘constexpr fmt::v9::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; Args = {unsigned int, std::basic_string_view<char, std::char_traits<char> >, NPT_String}]’:
/usr/include/fmt/core.h:1918:31:   required from ‘constexpr fmt::v9::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<Args>::type>::type ...> fmt::v9::make_format_args(Args&& ...) [with Context = basic_format_context<appender, char>; Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:370:68:   required from ‘void spdlog::logger::log_(spdlog::source_loc, spdlog::level::level_enum, spdlog::string_view_t, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; spdlog::string_view_t = fmt::v9::basic_string_view<char>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:90:13:   required from ‘void spdlog::logger::log(spdlog::source_loc, spdlog::level::level_enum, fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:96:12:   required from ‘void spdlog::logger::log(spdlog::level::level_enum, fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/kodi-build/build/include/spdlog/logger.h:158:12:   required from ‘void spdlog::logger::info(fmt::v9::format_string<T ...>, Args&& ...) [with Args = {unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&}; fmt::v9::format_string<T ...> = fmt::v9::basic_format_string<char, unsigned int&, std::basic_string_view<char, std::char_traits<char> >&, const NPT_String&>]’
/home/ihipop/.cache/yay/kodi-git/src/xbmc/xbmc/network/upnp/UPnPInternal.cpp:1187:21:   required from here
/usr/include/fmt/core.h:1899:9: 错误：no matching function for call to ‘fmt::v9::detail::arg_data<fmt::v9::detail::value<fmt::v9::basic_format_context<fmt::v9::appender, char> >, char, 3, 0>::arg_data(<brace-enclosed initializer list>)’
 1899 |         data_{detail::make_arg<
      |         ^~~~~~~~~~~~~~~~~~~~~~~
 1900 |             is_packed, Context,
      |             ~~~~~~~~~~~~~~~~~~~
 1901 |             detail::mapped_type_constant<remove_cvref_t<T>, Context>::value>(
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1902 |             FMT_FORWARD(args))...} {
      |             ~~~~~~~~~~~~~~~~~~~~~~
/usr/include/fmt/core.h:1169:28: 附注：candidate: ‘template<class ... U> constexpr fmt::v9::detail::arg_data<T, Char, NUM_ARGS, 0>::arg_data(const U& ...) [with U = {U ...}; T = fmt::v9::detail::value<fmt::v9::basic_format_context<fmt::v9::appender, char> >; Char = char; long unsigned int NUM_ARGS = 3]’
 1169 |   FMT_CONSTEXPR FMT_INLINE arg_data(const U&... init) : args_{init...} {}
      |                            ^~~~~~~~
/usr/include/fmt/core.h:1169:28: 附注：  template argument deduction/substitution failed:
/usr/include/fmt/core.h:1164:8: 附注：candidate: ‘constexpr fmt::v9::detail::arg_data<fmt::v9::detail::value<fmt::v9::basic_format_context<fmt::v9::appender, char> >, char, 3, 0>::arg_data(const fmt::v9::detail::arg_data<fmt::v9::detail::value<fmt::v9::basic_format_context<fmt::v9::appender, char> >, char, 3, 0>&)’
 1164 | struct arg_data<T, Char, NUM_ARGS, 0> {
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/fmt/core.h:1164:8: 附注：  conversion of argument 1 would be ill-formed:
/usr/include/fmt/core.h:1164:8: 附注：candidate: ‘constexpr fmt::v9::detail::arg_data<fmt::v9::detail::value<fmt::v9::basic_format_context<fmt::v9::appender, char> >, char, 3, 0>::arg_data(fmt::v9::detail::arg_data<fmt::v9::detail::value<fmt::v9::basic_format_context<fmt::v9::appender, char> >, char, 3, 0>&&)’
/usr/include/fmt/core.h:1164:8: 附注：  conversion of argument 1 would be ill-formed:
make[2]: *** [build/network/upnp/CMakeFiles/network_upnp.dir/build.make:90：build/network/upnp/CMakeFiles/network_upnp.dir/UPnPInternal.cpp.o] 错误 1
make[1]: *** [CMakeFiles/Makefile2:13960：build/network/upnp/CMakeFiles/network_upnp.dir/all] 错误 2
make: *** [Makefile:146：all] 错误 2
==> 错误： 在 build() 中发生一个错误。
    正在放弃...
```

```
使用内建 specs。
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.0/lto-wrapper
目标：x86_64-pc-linux-gnu
配置为：/build/gcc/src/gcc/configure --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --enable-bootstrap --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --with-build-config=bootstrap-lto --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-libstdcxx-backtrace --enable-link-serialization=1 --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-werror
线程模型：posix
Supported LTO compression algorithms: zlib zstd
gcc 版本 12.2.0 (GCC) 
```
```
cmake --version       
cmake version 3.24.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

Reply:
I removed the  the cast and we will see if Jenkins build failed too
if failed, I will revert the change

Reply:
@enen92 this type cast is required,Jenkins build failed too with same error on my machine

https://jenkins.kodi.tv/job/LINUX/46491/console

Reply:
its because its a wrapped object to avoid the cast you should use `m_Uri.GetChars()`

Reply:
@CastagnaIT  Thanks .